# Memo Pandas

A travers ce notebook nous verons les fonctions les plus utiles de panda. <br>
La librarie Pandas est très utilisé pour la création, la manipulation et la visualisation de dataframe avec python.

In [0]:
import pandas as pd

## Importation de données

Il est possible de créer un dataframe à partir d'objat pyhton.

In [2]:
# Création d'un dataframe à partir d'une liste
data = [['prenom1', 21], ['prenom2', 32], ['prenom3', 12]] 
pd.DataFrame(data, columns = ['prenom', 'age'])

# Même Chose avec un dictionnaire
data = {'prenom':['prenom1', 'prenom2', 'prenom3'], 'age':[21, 32, 12]}
pd.DataFrame(data)

prenom  age
0  prenom1   21
1  prenom2   32
2  prenom3   12

Mais la plupart du temps nous utiliserons des données stocké dans un fichier externe. 
Pour cela nous pouvons utliser les fonctions de pandas qui permettent d'importer de nombreux types de fichiers et de les convertir en object pandas.
Voici un resumé des principales fonctions de pandas pour l'importation de données 

|   Type Fichier   |   Fonction |   Decription |   lien doc |
|:-:|:-:|:-:|:-:|
|Pickling|read_pickle|Load pickled pandas object (or any object) from file.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_pickle.html#pandas.read_pickle)|
|Flat file|read_table|Read general delimited file into DataFrame.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_table.html#pandas.read_table)|
|Flat file|read_csv|Read a comma-separated values (csv) file into DataFrame.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html#pandas.read_csv)|
|Flat file|read_fwf|Read a table of fixed-width formatted lines into DataFrame.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_fwf.html#pandas.read_fwf)|
|Flat file|read_msgpack|(DEPRECATED) Load msgpack pandas object from the specified file path.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_msgpack.html#pandas.read_msgpack)|
|Clipboard|read_clipboard|Read text from clipboard and pass to read_csv.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_clipboard.html#pandas.read_clipboard)|
|Excel|read_excel|Read an Excel file into a pandas DataFrame.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_excel.html#pandas.read_excel)|
|JSON|read_json|Convert a JSON string to pandas object.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_json.html#pandas.read_json)|
|HTML|read_html|Read HTML tables into a list of DataFrame objects.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_html.html#pandas.read_html)|
|HDF5|read_hdf|Read from the store, close it if we opened it.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_hdf.html#pandas.read_hdf)|
|Feather|read_feather|Load a feather-format object from the file path.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_feather.html#pandas.read_feather)|
|Parquet|read_parquet|Load a parquet object from the file path, returning a DataFrame.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_parquet.html#pandas.read_parquet)|
|SAS|read_sas|Read SAS files stored as either XPORT or SAS7BDAT format files.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_sas.html#pandas.read_sas)|
|SQL|read_sql_table|Read SQL database table into a DataFrame.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_sql_table.html#pandas.read_sql_table)|
|SQL|read_sql_query|Read SQL query into a DataFrame.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_sql_query.html#pandas.read_sql_query)|
|SQL|read_sql|Read SQL query or database table into a DataFrame.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_sql.html#pandas.read_sql)|
|Google BigQuery|read_gbq|Load data from Google BigQuery.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_gbq.html#pandas.read_gbq)|
|STATA|read_stata|Read Stata file into DataFrame.|[lien](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_stata.html#pandas.read_stata)|

Pour la suite, nous allons utiliser un jeux de données sur les maladies cardiaque disponible sur Kaggle.
Le jeux de données contient 13 variables :
* âge
* sexe
* type de douleur thoracique (4 valeurs)
* tension artérielle au repos
* cholestoral sérique en mg / dl
* glycémie à jeun> 120 mg / dl
* résultats électrocardiographiques au repos (valeurs 0,1,2)
* fréquence cardiaque maximale atteinte
* angine de poitrine
* pente du segment ST d'exercice maximal
* nombre de principaux vaisseaux (0-3) colorés par une fluoroscopie
* thal: 3 = normal; 6 = défaut fixe; 7 = défaut réversible

Nous avons rajouter des lignes à ce jeux de données pour illustrer quelques exemples <br>

Pour importer ce jeux de données, nous allons utiliser la fonction read_csv

In [0]:
data = pd.read_csv('data/heart.csv')

## Information sur le jeux de données

Une fois notre jeux de données importé, nous pouvons obtenir des informations sur celui ci. <br>
Dans un premier temps, nous pouvons visualiser un echantillon de données grace à la méthode **head()**

In [5]:
data.head()

age  sex   cp  trestbps  chol  fbs  ...  exang  oldpeak  slope   ca  thal  target
0   63  1.0  3.0     145.0   233  1.0  ...    0.0      2.3    0.0  0.0   1.0     1.0
1   37  1.0  2.0     130.0   250  0.0  ...    0.0      3.5    0.0  0.0   2.0     1.0
2   41  0.0  1.0     130.0   204  0.0  ...    0.0      1.4    2.0  0.0   2.0     1.0
3   56  1.0  1.0     120.0   236  0.0  ...    0.0      0.8    2.0  0.0   2.0     1.0
4   57  0.0  0.0     120.0   354  0.0  ...    1.0      0.6    2.0  0.0   2.0     1.0

[5 rows x 14 columns]

Nous allons utiliser seuleument quelques variables pour illustrer les exemples. Nous allons concerver uniquement les variables age, sex, chol et target. <br>
La variable target indique la présence d'une maladie cardiaque chez le patient

In [0]:
data = data[['age','sex','chol','target']]

Pour obtenir des information, sur les variables. Nous utilisons la méthode **info()**

In [7]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 304 entries, 0 to 303
Data columns (total 4 columns):
age       304 non-null int64
sex       303 non-null float64
chol      304 non-null int64
target    303 non-null float64
dtypes: float64(2), int64(2)
memory usage: 9.6 KB


Cette méthode permet de voir que notre jeux de données est composé de 304 lignes et de 4 varibales. <br>
La plupard des variables sont de type numerique acec 13 variables de type int et 1 de type float

Nous pouvons aussi visualiser le nombre de valeurs distinct pour chacune des variables

In [8]:
data.nunique()

age        41
sex         2
chol      152
target      2
dtype: int64

Pour obtenir des informations sur la distribution des variables, nous pouvons utiliser les méthodes suivantes, bien que cela ne veux rien dire sur les variables sexe et targets puisse qu'il sagit de varibales qualitatives

In [9]:
# Description du jeux de données
data.describe()

age         sex        chol      target
count  304.000000  303.000000  304.000000  303.000000
mean    54.401316    0.683168  246.263158    0.544554
std      9.087590    0.466011   51.745153    0.498835
min     29.000000    0.000000  126.000000    0.000000
25%     47.750000    0.000000  211.000000    0.000000
50%     55.500000    1.000000  240.500000    1.000000
75%     61.000000    1.000000  274.250000    1.000000
max     77.000000    1.000000  564.000000    1.000000

Ou encore

In [11]:
print(data['age'].max())
print(data['age'].min())
print(data['age'].mean())
print(data['age'].median())
print(data['age'].count())
print(data['age'].std())

77
29
54.401315789473685
55.5
304
9.087589817549375


Pour obtenir la matrice des correlations, nous pouvons utiliser la méthode corr() qui utilise la valeur du test de pearson par default. C'est pour cette raison que nous sélectionnons uniquement nos variables quantitative.  

In [12]:
data[['age','chol']].corr()

age      chol
age   1.000000  0.213177
chol  0.213177  1.000000

Nous pouvons aussi réaliser des agrégations comme dans l'exemple suivant

In [13]:
data['target'].value_counts()

1.0    165
0.0    138
Name: target, dtype: int64

On peut verifier la présence de valeur manqauntes

In [14]:
data.isnull().sum()

age       0
sex       1
chol      0
target    1
dtype: int64

## Manipulation de données

Suppression des valeurs manquantes

In [0]:
data.dropna(inplace=True)

Nous aurions aussi pu remplacer les valeurs manqauntes

In [0]:
data['age'] = data['age'].fillna(data['age'].mean())

Les variables sex et target ne contiennent que 2 modalités codé sous en entier. Nous allons transfomer ces deux variables pour qu'elle soit plus parlante

In [17]:
def rec_sex(valeur):
    if valeur == 1:
        return("Homme")
    else :
        return("Femme")
    
data['sex'] = data.sex.apply(rec_sex)

data.head()

age    sex  chol  target
0   63  Homme   233     1.0
1   37  Homme   250     1.0
2   41  Femme   204     1.0
3   56  Homme   236     1.0
4   57  Femme   354     1.0

Pour changer le format d'une variable

In [0]:
data['target'] = data.target.astype(int)

On peut concaténer des fichiers en ligne ou en colonne

In [19]:
data_concat = pd.concat([data,data], axis = 0)
print("shape = ",data_concat.shape)
data_concat.head()

shape =  (606, 4)


age    sex  chol  target
0   63  Homme   233       1
1   37  Homme   250       1
2   41  Femme   204       1
3   56  Homme   236       1
4   57  Femme   354       1

In [20]:
data_concat = pd.concat([data,data], axis = 1)
print("shape = ",data_concat.shape)
data_concat.head()

shape =  (303, 8)


age    sex  chol  target  age    sex  chol  target
0   63  Homme   233       1   63  Homme   233       1
1   37  Homme   250       1   37  Homme   250       1
2   41  Femme   204       1   41  Femme   204       1
3   56  Homme   236       1   56  Homme   236       1
4   57  Femme   354       1   57  Femme   354       1

On peut réaliser des jointures entre dataframe par exemple nous allons remplacer la variable target par "présence" ou "abscence" avec une jointure

In [21]:
data_merge = pd.DataFrame([[1, 'présence'], [0, 'abscence']], columns = ['target', 'maladie_cardiaque'])

data = pd.merge(data,data_merge, how='inner', on='target')
data.head()

age    sex  chol  target maladie_cardiaque
0   63  Homme   233       1          présence
1   37  Homme   250       1          présence
2   41  Femme   204       1          présence
3   56  Homme   236       1          présence
4   57  Femme   354       1          présence

Pour renommer des variables

In [22]:
data.rename(columns={'chol':'tx_cholesterol'}, inplace = True)
data.head()

age    sex  tx_cholesterol  target maladie_cardiaque
0   63  Homme             233       1          présence
1   37  Homme             250       1          présence
2   41  Femme             204       1          présence
3   56  Homme             236       1          présence
4   57  Femme             354       1          présence

## Agrégation

On peut réaliser differents types d'aggrégations avec pandas.

In [34]:
# Pour avoir la Fréquence des modalités d'une variable
data.maladie_cardiaque.value_counts(normalize = True)

présence    0.544554
abscence    0.455446
Name: maladie_cardiaque, dtype: float64

In [28]:
# Moyenne par groupes
data.groupby(['maladie_cardiaque']).mean()

age  tx_cholesterol  target
maladie_cardiaque                                   
abscence           56.601449      251.086957     0.0
présence           52.496970      242.230303     1.0

In [29]:
# Moyenne par groupes
data.groupby(['sex','target']).mean()

age  tx_cholesterol
sex   target                           
Femme 0       59.041667      274.958333
      1       54.555556      256.750000
Homme 0       56.087719      246.061404
      1       50.903226      230.989247

In [30]:
# Trier une dataframe
data.sort_values('age').head()

age    sex  tx_cholesterol  target maladie_cardiaque
72    29  Homme             204       1          présence
58    34  Homme             182       1          présence
125   34  Femme             210       1          présence
239   35  Homme             282       0          abscence
65    35  Femme             183       1          présence

Une fois notre prétairement réalisé, nous pouvons exporter notre table de résulats dans differents formats. <br>
Voici un exemple d'export en csv.

In [0]:
data[['age','sex','tx_cholesterol','maladie_cardiaque']].to_csv('data/data_clean.csv',index = False)

### Autres fonctions utiles

In [0]:
# Transformer un nombre de seconde en datetime :
pd.to_datetime(variable_second, unit='s')